src/OpenDict.h

```diff
@@ -0,0 +1,351 @@
+#endif //LOAD_FACTOR_BITS
+
+//when log2_buckets > DICT_THRESHOLD_BITS, DICT_LOAD_FACTOR_BITS becomes effective.
+//basicly it means if dict size < 2^DICT_THRESHOLD_BITS+n, size up only if necessary (the last entry can't hold new key).
+#ifndef DICT_THRESHOLD_BITS
+#define DICT_THRESHOLD_BITS 3
+#endif
+
+// of insertions.
+typedef enum { ORDERED, UNORDERED } dict_order;
+
+// Type for function to be called when deleting elements.
+typedef void (*dict_delete_func)(void*);
+
+// A dict_delete_func that just calls delete.
+extern void generic_delete_func(void*);
+// The value of an iteration cookie is the bucket and offset within the
+// bucket at which to start looking for the next value to return.
+#define TOO_FAR_TO_REACH 0xFFFF
+#define MAX_KEY_SIZE 0XFFFF
+struct DictEntry{//24 bytes. perfectly alligned. always own the key. but not the value.
+	#ifdef DEBUG
+	int bucket;//for easy debugging
+	#endif//DEBUG
+	uint16_t distance; //<from expected position. all 1's means empty. max distance 64K-2
+	uint16_t key_size; //the length of the key. <=8 will be embedded directly. otherwise a pointer 8K max enough?
+	uint32_t hash; //lower 4-byte of the 8-byte long hash. the part to calculate position in the table.
+
+	void* value; //for Bro, value is always a pointer. 
+	union{
+		char key_here[8]; //hold key len<=8. when over 8, it's a pointer to real keys.
+		char* key;
+	};
+
+	DictEntry(void* arg_key, int key_size=0, hash_t hash=0, void* value=0, int16_t d=TOO_FAR_TO_REACH, int copy_key=0) 
+	: distance(d), key_size(key_size), hash((uint32_t)hash), value(value)
+		{
+		#ifdef DEBUG
+		bucket = 0;
+		#endif//DEBUG
+		if( key_size <= 8)
+			{				
+			memcpy(key_here, arg_key, key_size);
+			if(!copy_key)
+				delete (char*)arg_key; //own the arg_key, now don't need it.
+			}
+		else
+			{
+			if( copy_key )
+				{
+				key = new char[key_size];
+				memcpy(key, arg_key, key_size);
+				}
+			else
+				{
+				key = (char*)arg_key;
+				}
+			}
+		}
+	bool Empty()	{ return distance == TOO_FAR_TO_REACH; }
+	void SetEmpty()
+		{
+		distance = TOO_FAR_TO_REACH;
+		#ifdef DEBUG
+		hash = 0;
+		key = NULL;
+		value = NULL;
+		key_size = 0;
+		bucket = 0;
+		#endif//DEBUG
+		}
+
+	//if with no intent to release key memory, call SetEmpty instead. key pointer is shared when moving around.
+	void Clear() 
+		{//SetEmpty & release memory if allocated.
+		if( key_size > 8 )
+			delete key;
+		SetEmpty();
+		}
+	const char* GetKey() const { return key_size <= 8? key_here : key; }
+	bool Equal(const char* arg_key, int arg_key_size, hash_t arg_hash) const
+		{//only 40-bit hash comparison.
+		return ( 0 == ((hash ^ arg_hash) & HASH_MASK) ) 
+			&& key_size == arg_key_size && 0 == memcmp(GetKey(), arg_key, key_size);
+		}
+	bool operator==(const DictEntry& r) const
+		{
+		return Equal(r.GetKey(), r.key_size, r.hash);
+		}
+	bool operator!=(const DictEntry& r) const
+		{
+		return ! Equal(r.GetKey(), r.key_size, r.hash);
+		}
+};
+
+struct IterCookie;
```

Comment:
There's already a forward declaration for this earlier in the file as a `class`. I think this can be removed.